accepted/2023/sign-signature-provider-plugins.md

```diff
@@ -0,0 +1,524 @@
+    /// <remarks>
+    /// The plugin host will try casting a <see cref="IPlugin" /> instance to this interface to determine if the
+    /// plugin is a signature provider.
+    /// </remarks>
+    public interface ISignatureProvider
+    {
+        Task<AsymmetricAlgorithm> GetAsymmetricAlgorithmAsync(CancellationToken cancellationToken);
+    }
+
+    /// <summary>
+    /// This is an optional interface for a signature provider plugin.
+    /// If the asymmetric key pair returned by <see cref="ISignatureProvider" /> has an associated X.509 certificate,
+    /// the plugin should also implement this interface.
+    /// <summary>
+    /// <remarks>
+    /// The plugin host will try casting a <see cref="ISignatureProvider" /> instance to this interface to determine
+    /// if the plugin is a certificate provider.  A plugin should not implement this interface unless it can provide
+    /// a certificate.  Do not implement this interface and throw <see cref="NotImplementedException" /> to indicate
+    /// non-support.
+    ///
+    /// Some signature formatters work with raw asymmetric key pairs without a certificate.  To support only these
+    /// signature formatters, a plugin need not implement this interface.
+    ///
+    /// Some signature formatters (like Authenticode) do require a certificate and will fail without a certificate.
+    /// To support these signature formatters, the plugin should implement this interface.  Signature formatters may
+    /// fail if they require a certificate but a signature provider plugin does not implement this interface.
+    /// </remarks>
+    public interface ICertificateProvider
+    {
+        Task<X509Certificate2> GetCertificateAsync(CancellationToken cancellationToken);
+    }
+}
+```
+
+Sign CLI will pass all plugin arguments as a read-only dictionary of name-value pairs.  The plugin is responsible for argument parsing and validation for all options defined by the plugin.
+
+The new interfaces assembly will be packaged and published to <https://nuget.org>, similar to [NuGetRecommender's contracts-only package](https://www.nuget.org/packages/Microsoft.DataAI.NuGetRecommender.Contracts).  The Sign CLI team will manage the source code repository for this package and publish the package to <https://nuget.org>.
+
+The interfaces package itself can have package dependencies; however, because Sign CLI and all plugins would inherit new interfaces package dependencies, we should exercise restraint and caution before adding new dependencies.  An example of a package dependency worth having is [`Microsoft.Extensions.Logging.Abstractions`](https://www.nuget.org/packages/Microsoft.Extensions.Logging.Abstractions).  Sign CLI uses [`ILogger`](https://learn.microsoft.com/dotnet/api/microsoft.extensions.logging.ilogger) ubiquitously, it makes sense for plugins to write to a shared logger.  From the provided `System.IServiceProvider` argument, plugins can request an `ILogger` instance for logging.
```

Comment:
> From the provided `System.IServiceProvider` argument

Example above uses concrete `ServiceProvider`.  The text should be aligned with the code sample.